base/abstractarray.jl

```diff
@@ -2506,6 +2506,180 @@ end
     Ai
 end
 
+"""
+    stack(arrays)
+
+Concatenates a collection of arrays, all the same size, into one higher-dimensional array.
+
+The first dimension(s) are those of the individual arrays, followed by those from the
+container. Thus the result has size `(size(first(arrays))..., size(arrays)...)`.
+
+See also [`cat`](@ref), [`eachcol`](@ref).
+
+!!! compat "Julia 1.8"
+    This function requires at least Julia 1.8.
+
+# Examples
+```jldoctest
+julia> vecs = [[1,2], [3,4], [5,6]]
+3-element Vector{Vector{Int64}}:
+ [1, 2]
+ [3, 4]
+ [5, 6]
+
+julia> mat = stack(vecs)
+2×3 Matrix{Int64}:
+ 1  3  5
+ 2  4  6
+
+julia> mat == reduce(hcat, vecs) == hcat(vecs...)
+true
+
+julia> mat == stack(eachcol(mat))
+true
+
+julia> vec(mat) == reduce(vcat, vecs) == vcat(vecs...)
+true
+
+julia> mats = (fill(i/2,3,4) for i in (1, 10, 100) if i>pi);
+
+julia> stack(mats)
+3×4×2 Array{Float64, 3}:
+[:, :, 1] =
+ 5.0  5.0  5.0  5.0
+ 5.0  5.0  5.0  5.0
+ 5.0  5.0  5.0  5.0
+
+[:, :, 2] =
+ 50.0  50.0  50.0  50.0
+ 50.0  50.0  50.0  50.0
+ 50.0  50.0  50.0  50.0
+
+julia> ans == cat(mats..., dims=3)
+true
+```
+"""
+stack(itr) = _stack_iter(IteratorSize(itr), itr)
+stack(A::AbstractArray{<:AbstractArray}) = _typed_stack(mapreduce(eltype, promote_type, A), A)
+
+"""
+    stack(f, args)
+
+Apply `f` to each element of `args`, and `stack` the result.
+
+See also [`mapslices`](@ref), [`mapreduce`](@ref).
+
+# Examples
+```jldoctest
+julia> stack("julia") do c
+         (c, c-32)
+       end
+2×5 Matrix{Char}:
+ 'j'  'u'  'l'  'i'  'a'
+ 'J'  'U'  'L'  'I'  'A'
```

Comment:
I added the `stack(f, iter)` method suggested.

Notice that this allows `f` to return tuples, and treats them just like vectors. This is not how `cat` behaves, but if it's a different verb then perhaps we have more freedom. The reason to want this is that, if your function always returns (say) 3-vectors, then this might let you avoid allocating many small vectors.

You could go further and make `stack([(1,2), (3,4)])` into `reinterpret(reshape, Int, [(1,2), (3,4)])`.

Reply:
However, this has type-stability problems right now?
```
julia> x = randn(3, 1000);

julia> @btime stack(eachcol($x));
  min 4.774 μs, mean 7.023 μs (4 allocations, 23.58 KiB. GC mean 12.95%)

julia> @btime stack(collect, eachcol($x));
  min 22.875 μs, mean 28.216 μs (1004 allocations, 101.70 KiB. GC mean 11.49%)

julia> @btime stack(NTuple{3,Float64}, eachcol($x));  # best case
  min 4.768 μs, mean 6.950 μs (4 allocations, 23.58 KiB. GC mean 12.92%)

julia> ft(xyz) = @inbounds (x[1], x[2]/2, x[3]^3);
julia> fv(xyz) = @inbounds [x[1], x[2]/2, x[3]^3];

julia> @btime stack(fv, eachcol($x));  # here type-stability is broken
  min 154.000 μs, mean 165.236 μs (6841 allocations, 192.95 KiB. GC mean 4.81%)

julia> @btime stack(ft, eachcol($x));  # tuples, can I save allocations?
  min 231.500 μs, mean 241.493 μs (7839 allocations, 177.31 KiB. GC mean 2.97%)
```